### ISSUE TITLE
Unstable routine name suffix generation with overloaded names

### ISSUE THREAD

**Issue by demichej** [jOOQ/jOOQ#9754]
### Your question:

We're hitting an issue with Postgres code generation and having 2 routines with the same name (but different arguments). I did notice this bug: https://github.com/jOOQ/jOOQ/issues/4970, but it appears that there can still be an inconsistency.

See the screenshot of the difference from one code generation to another.

<img width="1608" alt="Screen Shot 2020-01-22 at 11 12 53 AM" src="https://user-images.githubusercontent.com/422315/72919011-4d0fbb80-3d0c-11ea-8c17-e42902b6d634.png">

Here are the routine defs
```
-[ RECORD 1 ]-------+----------------------------------------------------------------------------
Schema              | public
Name                | sync_update_table
Result data type    | void
Argument data types | t_name character varying
Type                | normal
-[ RECORD 2 ]-------+----------------------------------------------------------------------------
Schema              | public
Name                | sync_update_table
Result data type    | void
Argument data types | t_name character varying, test_drive boolean, record_schema_changes boolean
Type                | normal
```
Maybe order on argument data types as well?

### Versions:

- jOOQ: 3.12
- Java: 1.8.0_231
- Database (include vendor): Postgres 10.7
- OS: Linux on one code generation and Mac on another
- JDBC Driver (include name if inofficial driver):42.1.1

**Comment by knutwannheden** [comment]
@demichej Thank you very much for your report. We will investigate the issue and report our findings back to you.

**Comment by knutwannheden** [comment]
@demichej Is it possible that you drop and recreate the routines between the code generation runs or that you run the code generator against different database instances? I am asking because this could in fact cause the issue you describe.

We will investigate whether we find some way to do the code generation in a way that is stable with regards to the signatures of the overloaded procedures. Of course, if the signature of any of the routines were to change, this would then again lead to a renumbering of the routines. While I think this would be a better strategy, it also has this type of problem.

As a workaround you should be able to implement a class extending `DefaultGeneratorStrategy` which overrides the method `getOverloadSuffix()`. In there you should be able to inspect the given `RoutineDefinition`'s details and implement your own fixed suffix mapping for your overloaded procedures.

**Comment by demichej** [comment]
Thanks for the response @knutwannheden. You're right, it was generated off of 2 different database instances. I'll take a look at overriding that method for code generation as well. Thanks!

**Comment by lukaseder** [comment]
Starting from jOOQ 3.13, we'll implement a default `<orderProvider/>` in the absence of a user provided one: https://github.com/jOOQ/jOOQ/issues/9604

Historically, we have defined the order of generated objects in `jOOQ-meta` SQL queries, with the occasional risk of a bug that has lead to such instability, in case the order was not stable. In the future, this should not longer be the case, as we'll order the `jOOQ-meta` objects again in memory according to a standard scheme for all dialects, which is no longer dependent on the dialect specific SQL queries. So, this should resolve itself with 3.13.

Until then, as a workaround, a custom `<orderProvider/>` can be specified, see:
https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-order-provider

**Comment by lukaseder** [comment]
My bad. This is not about sorting the objects, but about the suffix generation, which then leads to different sorting, of course. The suffix is generated as follows:

```java
when(
    count().over(partitionBy(r1.ROUTINE_SCHEMA, r1.ROUTINE_NAME)).gt(one()),
    rowNumber().over(partitionBy(r1.ROUTINE_SCHEMA, r1.ROUTINE_NAME).orderBy(r1.SPECIFIC_NAME))
).as("overload"),
```

It seems indeed that our approach is incorrect as it depends on the specific database instance that is used to generate code. We should probably switch to using the routine signature instead. Looks like this turns into a backwards incompatible change...

**Comment by lukaseder** [comment]
We actually seem to have this problem in our integration tests, just never addressed it. We generate code for `regexp_matches` and the overload order keeps changing depending on who is generating it.

Will look into this

**Comment by lukaseder** [comment]
A first attempt is to calculate the type signature with a correlated subquery embedded in the window function's partition by clause:

```java
when(
    count().over(partitionBy(r1.ROUTINE_SCHEMA, r1.ROUTINE_NAME)).gt(one()),
    rowNumber().over(partitionBy(r1.ROUTINE_SCHEMA, r1.ROUTINE_NAME).orderBy(

        // [#9754] To stabilise overload calculation, we use the type signature
        field(
            select(arrayAgg(p.DATA_TYPE).orderBy(p.ORDINAL_POSITION))
            .from(p)
            .where(p.SPECIFIC_NAME.eq(r1.SPECIFIC_NAME))
        ),
        r1.SPECIFIC_NAME
    ))
).as("overload"),
```

Not entirely surprisingly, this doesn't seem to perform. Will try other things...

**Comment by lukaseder** [comment]
`pg_proc` may hold the answer, as the argument types are already there:

```sql
select proname, proargtypes
from pg_proc
order by proname, proargtypes
```

E.g.:

```
array_to_json	{2277}
array_to_json	{2277,16}
```

Looking up type names might be tedious, and IDs are equally unreliable

**Comment by lukaseder** [comment]
Bingo:

```sql
select proname, replace(pg_get_function_arguments(pg_proc.oid), '"', '')
from pg_proc
order by 1, 2;
```

Yielding:

```
array_to_json	anyarray
array_to_json	anyarray, boolean
```

**Comment by lukaseder** [comment]
While this fix is desireable, it may break a lot of existing generated code. I'm not too happy with that. The existing order is unstable in some cases, but when custom functions are installed in the same order, then the "specific name" order should reflect that, so the instability doesn't happen all the time.

Perhaps, we need a more global overload naming strategy, also one that is consistent across dialects. I prefer not to rush this fix.

**Comment by demichej** [comment]
Agreed that it's not desirable to really break a lot of generated code. We're ok on our end for now. I am simply overriding the `getOverloadSuffix()` as suggested, and I'm using the number of argument parameters and am using that instead of the default `overloadIndex`. That works for us at the moment. It makes the generation consistent across our databases.

**Comment by lukaseder** [comment]
This seems to be the same issue as #8589, which has been fixed for jOOQ 3.14.0 and 3.13.5 (#10533)

**Comment by lukaseder** [comment]
I've reopened https://github.com/jOOQ/jOOQ/issues/8589, as there still seem to be instabilities as of jOOQ 3.17
